### PR TITLE
ServiceAccountName replaces ServiceAccount

### LINKED ISSUE #1219: Change serviceAccount to serviceAccountName to be consistent with k8s
The `serviceAccount` field name should probably be renamed to `serviceAccountName` to match more closely with the normal k8s API. See `serviceAccountName` under pod [here](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.12/). 

This stems from @bobcatfish 's comment here: https://github.com/tektoncd/triggers/pull/63/files#r314863023

> ugh but it's called `serviceAccountName` in the rest of kubernetes isn't it :S
> 
> image
> 
> maybe we should leave this and change pipelines after all

### PR DESCRIPTION
**Pull request by fraenkel**
# Changes
Following in the k8s footsteps, deprecate ServiceAccount and replace it
with ServiceAccountName.

ServiceAccountName will always take precedence over ServiceAccount. If
ServiceAccountName is not set, the value provided by ServiceAccount will
be used instead.

Fixes #1219 

Before I go and fix all the yamls and documentation, I wanted to get a sense of the approach. I tried to find an optimal way but alas, given the way `ObjectDefaulter` and `SetDefaults` work, neither could accomplish the switch properly. `ObjectDefaulter` could have done it, but the way the tests are configured do not cause the Defaults to be executed since the objects aren't actually decoded.

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [x] Includes [tests](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if functionality changed/added)
- [x] Includes [docs](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if user facing)
- [x] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/master/standards.md#commit-messages)

## Reviewer Notes

If [API changes](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md)
are included, [additive changes](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md#additive-changes)
must be approved by at least two [OWNERS](https://github.com/tektoncd/pipeline/blob/master/OWNERS)
and [backwards incompatible changes](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md#backwards-incompatible-changes)
must be approved by [more than 50% of the OWNERS](https://github.com/tektoncd/pipeline/blob/master/OWNERS),
and they must first be added
[in a backwards compatible way](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md#backwards-compatible-changes-first).

# Release Notes

```
Describe any user facing changes here, or delete this block.

Examples of user facing changes:
- API changes
- Bug fixes
- Any changes in behavior

```

### CODE REVIEW

**Comment by afrittoli** [comment]
Thanks for this! 
The code looks good; based on your commit message, I would like to dig a bit more into the `ObjectDefaulter` to understand if there is anything we should change in the tests so that using that becomes a viable option.
I think it would be good to have more eyes on this as I feel we may be doing more of this in the future, especially since we're aiming at getting our APIs to beta.

**Comment by fraenkel** [comment]
The ObjectDefaulter is only called in 2 places that I know of which is just at the beginning of reconcile. While this works for the objects you see right then, it doesn't work for the ones that are obtained via the Lister. I was trying to find some common path between running code and local testing but so far I haven't.

File: pkg/apis/pipeline/v1alpha1/taskrun_types.go [hunk 1]
```diff
@@ -306,3 +310,11 @@ func (tr *TaskRun) IsCancelled() bool {
 func (tr *TaskRun) GetRunKey() string {
 	return fmt.Sprintf("%s/%s/%s", "TaskRun", tr.Namespace, tr.Name)
 }
+
+func GetTaskRunServiceAccountName(taskRun *TaskRun) string {
```
Comment (vdemeester) [inline pkg/apis/pipeline/v1alpha1/taskrun_types.go]: Any reason for this over `func (tr *TaskRun) GetServiceAccountName() string {` ? (in the end it's the same but we have others *attached* functions already on `TaskRun` :stuck_out_tongue:)

Comment (fraenkel) [reply pkg/apis/pipeline/v1alpha1/taskrun_types.go]: It started elsewhere and migrated here to satisfy a common location.
I wasn't sure how these "helper" methods wanted to live.

**Comment by afrittoli** [comment]
@fraenkel If you rebase this on Monday or Tuesday, it could still make it into the 0.7 release.

**Comment by fraenkel** [comment]
The failed integration test confuses me because the pipeline that it cannot find is listed right above it.

**Comment by vdemeester** [comment]
```
init_test.go:113: Create namespace arendelle-n2lms to deploy to
init_test.go:125: Verify SA "default" is created in namespace "arendelle-n2lms"
pipelinerun_test.go:155: Setting up test resources for "service account propagation and pipeline param" test in namespace arendelle-n2lms
pipelinerun_test.go:164: Waiting for PipelineRun pipelinerun2 in namespace arendelle-n2lms to complete
pipelinerun_test.go:166: Error waiting for PipelineRun pipelinerun2 to finish: pipeline run pipelinerun2 failed!
artifact_bucket_test.go:56: ############################################
artifact_bucket_test.go:56: ### Dumping objects from arendelle-n2lms ###
artifact_bucket_test.go:56: ############################################
asm_amd64.s:522: 
    ---
    apiVersion: tekton.dev/v1alpha1
    kind: Pipeline
    metadata:
      creationTimestamp: "2019-09-16T11:28:57Z"
      generation: 1
      name: pipeline2
      namespace: arendelle-n2lms
      resourceVersion: "5211"
      selfLink: /apis/tekton.dev/v1alpha1/namespaces/arendelle-n2lms/pipelines/pipeline2
      uid: 2bfc56ff-d875-11e9-88ba-42010a8001b2
    spec:
      params:
      - name: path
        type: string
      - name: dest
        type: string
      tasks:
      - name: task1
        params:
        - name: path
          value: $(params.path)
        - name: dest
          value: $(params.dest)
        taskRef:
          name: task2
    status: {}
    
    ---
    apiVersion: tekton.dev/v1alpha1
    kind: PipelineRun
    metadata:
      creationTimestamp: "2019-09-16T11:28:57Z"
      generation: 1
      labels:
        hello-world-key: hello-world-value
      name: pipelinerun2
      namespace: arendelle-n2lms
      resourceVersion: "5213"
      selfLink: /apis/tekton.dev/v1alpha1/namespaces/arendelle-n2lms/pipelineruns/pipelinerun2
      uid: 2bfe536c-d875-11e9-88ba-42010a8001b2
    spec:
      params:
      - name: path
        value: docker://gcr.io/build-crd-testing/secret-sauce
      - name: dest
        value: dir:///tmp/
      pipelineRef:
        name: pipeline2
      podTemplate: {}
      serviceAccountName: service-account2
      timeout: 1h0m0s
    status:
      completionTime: "2019-09-16T11:28:57Z"
      conditions:
      - lastTransitionTime: "2019-09-16T11:28:57Z"
        message: Pipeline arendelle-n2lms/pipeline2 can't be found:pipeline.tekton.dev
          "pipeline2" not found
        reason: CouldntGetPipeline
        status: "False"
        type: Succeeded
      startTime: "2019-09-16T11:28:57Z"
    
    ---
    apiVersion: tekton.dev/v1alpha1
    kind: Task
    metadata:
      creationTimestamp: "2019-09-16T11:28:57Z"
      generation: 1
      name: task2
      namespace: arendelle-n2lms
      resourceVersion: "5209"
      selfLink: /apis/tekton.dev/v1alpha1/namespaces/arendelle-n2lms/tasks/task2
      uid: 2bf9acf2-d875-11e9-88ba-42010a8001b2
    spec:
      inputs:
        params:
        - name: path
          type: string
        - name: dest
          type: string
      steps:
      - args:
        - copy
        - $(inputs.params.path)
        - $(inputs.params.dest)
        command:
        - skopeo
        image: quay.io/rhpipeline/skopeo:alpine
        name: config-docker
        resources: {}
    
artifact_bucket_test.go:56: #####################################################
artifact_bucket_test.go:56: ### Dumping logs from Pods in the arendelle-n2lms ###
artifact_bucket_test.go:56: #####################################################
asm_amd64.s:522: Deleting namespace arendelle-n2lms
```

Interesting, it looks like a race…

File: pkg/apis/pipeline/v1alpha1/pipelinerun_types.go [hunk 2]
```diff
@@ -50,7 +50,11 @@ type PipelineRunSpec struct {
...
 	Params []Param `json:"params,omitempty"`
 	// +optional
-	ServiceAccount string `json:"serviceAccount"`
+	ServiceAccountName string `json:"serviceAccountName,omitempty"`
+	// DeprecatedServiceAccount is a depreciated alias for ServiceAccountName.
+	// Deprecated: Use serviceAccountName instead.
+	// +optional
+	DeprecatedServiceAccount string `json:"serviceAccount,omitempty"`
 	// +optional
 	ServiceAccounts []PipelineRunSpecServiceAccount `json:"serviceAccounts,omitempty"`
```
Comment (vdemeester) [inline pkg/apis/pipeline/v1alpha1/pipelinerun_types.go]: Should those be `ServiceAccountNames` ? :stuck_out_tongue_closed_eyes:

Comment (fraenkel) [reply pkg/apis/pipeline/v1alpha1/pipelinerun_types.go]: I couldn't decide if it was similar to ServiceAccountName -> ServiceAccount in v1.core or not.
Let me know if you want it switched.

**Comment by sbwsg** [comment]
I don't see why the coverage report for pipelinerun_types.go has dropped so much but it looks like `GetServiceAccountName` could use some unit tests.  Apart from that I think this PR is ready to go.

File: docs/pipelineruns.md [hunk 3]
```diff
@@ -38,11 +38,16 @@ following fields:
...
+  - [`serviceAccountName`](#service-account) - Specifies a `ServiceAccount` resource
     object that enables your build to run with the defined authentication
+<<<<<<< HEAD
     information. When a `ServiceAccount` isn't specified, the `default-service-account`
     specified in the configmap - config-defaults will be applied.
   - [`serviceAccounts`](#service-accounts) - Specifies a list of `ServiceAccount`
+=======
+    information.
+  - [`serviceAccountNames`](#service-accounts) - Specifies a list of `ServiceAccountName`
+>>>>>>>  ServiceAccountNames replaces ServiceAccounts
```
Comment (sbwsg) [inline docs/pipelineruns.md]: looks like a git merge artifact might have snuck in.

File: pkg/apis/pipeline/v1alpha1/pipelinerun_types.go [hunk 4]
```diff
@@ -55,9 +55,14 @@ type PipelineRunSpec struct {
...
-	ServiceAccount string `json:"serviceAccount"`
+	ServiceAccountName string `json:"serviceAccountName,omitempty"`
+	// DeprecatedServiceAccount is a depreciated alias for ServiceAccountName.
+	// Deprecated: Use serviceAccountName instead.
 	// +optional
-	ServiceAccounts []PipelineRunSpecServiceAccount `json:"serviceAccounts,omitempty"`
+	DeprecatedServiceAccount string `json:"serviceAccount,omitempty"`
+	// +optional
+	DeprecatedServiceAccounts []DeprecatedPipelineRunSpecServiceAccount `json:"serviceAccounts,omitempty"`
+	ServiceAccountNames       []PipelineRunSpecServiceAccountName       `json:"serviceAccountNames,omitempty"`
```
Comment (sbwsg) [inline pkg/apis/pipeline/v1alpha1/pipelinerun_types.go]: Should `ServiceAccountNames` also be marked `// +optional`?

File: docs/pipelineruns.md (hunk 3 above)
Comment (fraenkel) [reply docs/pipelineruns.md]: fixed

File: pkg/apis/pipeline/v1alpha1/pipelinerun_types.go (hunk 4 above)
Comment (fraenkel) [reply pkg/apis/pipeline/v1alpha1/pipelinerun_types.go]: yup.

**Comment by fraenkel** [comment]
> I don't see why the coverage report for pipelinerun_types.go has dropped so much but it looks like `GetServiceAccountName` could use some unit tests. Apart from that I think this PR is ready to go.

The coverage accounting is a bit incorrect. These methods are fully tested via testcases in other packages but the default coverage will only calculate within the package. There are tools to accumulate coverage across packages for a more accurate picture.

**Comment by bobcatfish** [comment]
Looks like we've got another conflict (sorry @fraenkel !! let us know if you want us to resolve it) but otherwise should be good to go (correct me if I'm wrong @sbwsg @afrittoli @vdemeester )

**Comment by fraenkel** [comment]
looks like a network issue
```
I1008 15:06:29.351]         {"level":"error","ts":1570547180.7409866,"logger":"fallback-logger","caller":"git/git.go:35","msg":"Error running git [fetch --depth=1 --recurse-submodules=yes origin c15aced]: exit status 128\nfatal: unable to access 'https://github.com/tektoncd/pipeline/': Could not resolve host: github.com\n","stacktrace":"github.com/tektoncd/pipeline/pkg/git.run\n\t/go/src/github.com/tektoncd/pipeline/pkg/git/git.go:35\ngithub.com/tektoncd/pipeline/pkg/git.Fetch\n\t/go/src/github.com/tektoncd/pipeline/pkg/git/git.go:86\nmain.main\n\t/go/src/github.com/tektoncd/pipeline/cmd/git-init/main.go:36\nruntime.main\n\t/usr/local/go/src/runtime/proc.go:201"}
```

**Comment by vdemeester** [comment]
@fraenkel @bobcatfish I think we need to create an issue to remove the deprecated field in the release after the next one :angel:

**Pull request merged**